source/MaterialXGenShader/TypeDesc.cpp

```diff
@@ -28,6 +28,13 @@ TypeDescNameMap& typeNameMap()
     return map;
 }
 
+using StructTypeDescStorage = vector<StructTypeDesc>;
+StructTypeDescStorage& structTypeStorage()
+{
+    static StructTypeDescStorage  storage;
```

Comment:
Ah, I hadn't noticed this singleton pattern until now, and this seems like a design decision that we should discuss.

Would this break in situations where MaterialX is leveraged across multiple threads, with each thread performing shader generation on a different document with different usage of structs?

Traditionally we forbid singletons in MaterialX to avoid this class of multi-threading hazards, and if there's another engineering solution that stores this data in a thread-local object such as the `GenContext` (which was explicitly designed to handle this category of storage), that would be strongly preferable.

Reply:
This use intentionally parallels the existing pattern for the TypeDesc system [here](https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/source/MaterialXGenShader/TypeDesc.cpp#L21)

```
TypeDescMap& typeMap()
{
    static TypeDescMap map;
    return map;
}
```

If we want to refactor the type system to use storage in `GenContext` I would propose that we do that as a separate PR.

Reply:
Got it, and it sounds we have some existing threading issues to address in our shader generation, independent of this proposal.  I'll close out this request for now, but I'm CC'ing @niklasharrysson so that he's in the loop.

Reply:
@jstone-lucasfilm @ld-kerley Just adding my 2 cents on the threading issue:

The use of a singleton for `typeMap()` does not expose threading issues, since write access to this data is strictly private to this file and only done once during program initialization. This is because the registration of TypeDesc's for native data types is hard coded and only done here:

https://github.com/AcademySoftwareFoundation/MaterialX/blob/main/source/MaterialXGenShader/TypeDesc.cpp#L118

After this we only do read access.

Whereas for StructTypeDesc's, the registration is dynamic based on the loaded documents, so in this case there are potential threading issues with using the same pattern.

If we at some point want to make registration of TypeDesc's data driven as well, we would need to update both of these singleton use cases though.